<p><img alt="Colaboratory logo" height="140px" src="https://upload.wikimedia.org/wikipedia/commons/archive/f/fb/20161010213812%21Escudo-UdeA.svg" align="left" hspace="10px" vspace="0px"></p>

# **Curso de bases para AI - Universidad de Antioquia**


El presente curso hace parte del centro de Big Data de la facultad de ciencias exactas y naturales (FCEN) de la Universidad de Antioquia.

\

El curso será dictado en 16 sesiones o clases de 4 horas los días sábado en las instalaciones de la universidad. Para un total de dededicación temporal de 64 horas. El curso cuenta con los siguientes módulos:

1. Bases matemáticas y de programación. (10H)
2. Bases de Python. (19H)
3. Bases de R. (13H)
4. Procesamiento de datos. (22H)

In [0]:
import pandas as pd
import numpy as np

# Lectura de datos en Pandas

El módulo de Pandas tiene una gran cantidad de funciones para la letura de datos en distintos formatos, tan solo vamos a mencionar algunas cosas sobre ellas. No olvide consultar constantemente la [documentación](https://pandas.pydata.org/docs/user_guide/io.html) para enterarse de las novedades del API.

**CSV** es el formato texto plano más usado en el mundo de los datos y con el cual vamos a trabajar en todo el curso. Sin embargo no es el único formato texto que se conoce, hay dos en particular: los formatos de [Excel](https://pandas.pydata.org/docs/user_guide/io.html#io-excel-reader) y los [JSON](https://pandas.pydata.org/docs/user_guide/io.html#io-json-reader). 


## Formato xls o xlsx

Para la lectura de estos datos tenemos: 

In [3]:
path_to_file = "https://github.com/diplomado-bigdata-machinelearning-udea/curso0/blob/master/s07/Tasa_de_Cambio_Representativa_del_Mercado-_TRM.xlsx?raw=True"
eneroTMR = pd.read_excel(path_to_file, sheet_name='Enero')
eneroTMR.head()

VALOR UNIDAD VIGENCIADESDE VIGENCIAHASTA
0  3,258.84    COP    2020-01-03    2020-01-03
1  3,262.05    COP    2020-01-04    2020-01-07
2  3,264.26    COP    2020-01-08    2020-01-08
3  3,254.42    COP    2020-01-09    2020-01-09
4  3,253.89    COP    2020-01-10    2020-01-10

In [4]:
enefebTMR = pd.read_excel(path_to_file, sheet_name=[0,1])
print(type(enefebTMR))
print(enefebTMR.keys())
enefebTMR[1].head()

<class 'dict'>
dict_keys([0, 1])


VALOR UNIDAD VIGENCIADESDE VIGENCIAHASTA
0  3,423.24    COP    2020-02-01    2020-02-03
1  3,401.56    COP    2020-02-04    2020-02-04
2  3,368.87    COP    2020-02-05    2020-02-05
3  3,355.44    COP    2020-02-06    2020-02-06
4  3,378.43    COP    2020-02-07    2020-02-07

In [5]:
TMR2020 = pd.read_excel(path_to_file,sheet_name=None)
print(TMR2020.keys())

dict_keys(['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo'])


## El formato JSON

JSON o JavaScript Object Notation es un formato de intercambio de Datos que se considera uno de los objetos más independientes del lenguaje. JSON surgió de la necesidad de un protocolo de comunicación de servidor a navegador en tiempo real y es uno de los formatos más extendido para tal fin. 

Si vamos a la [página](https://www.datos.gov.co/browse?sortBy=newest) oficial de datos de Colombia, tendremos una enorme base de datos para trabajar que podemos leer en formato JSON. 

Vamos a leer los [datos]("https://www.datos.gov.co/Transporte/Informe-de-carga-a-rea/4wwa-qb9a") de carga áerea en medio del aislamiento. En la página usted puede descargar los datos de dos maneras:

![texto alternativo](https://github.com/diplomado-bigdata-machinelearning-udea/curso0/blob/master/s07/imagenes/read.png?raw=true)

Cuando usted entra a **API**, allí le habla un poco sobre lo que es SODA, Socrata Open Data API, puede revisar sobre este tipo de API para manejo de bases abiertas, pero en la parte inferrio hay un link de internet, seguido de dos opciones, escogemos JSON y copiamos el link: 


In [6]:
source = "https://www.datos.gov.co/resource/4wwa-qb9a.json"
info_carga = pd.read_json(source)
print(info_carga.shape)
info_carga.head()

(1000, 17)


fecha  ... carga_vuelos_internacionales
0  2020-03-25T00:00:00.000  ...                          353
1  2020-03-25T00:00:00.000  ...                            0
2  2020-03-25T00:00:00.000  ...                            0
3  2020-03-25T00:00:00.000  ...                            0
4  2020-03-25T00:00:00.000  ...                            0

[5 rows x 17 columns]

Es importante, y si revisamos la página, nos daremos cuenta que sólo nos ha dado 1000 filas, cuando la base de datos tiene más de 2000. Esto se debe a que las peticiones, es decir abrir el link, tiene un limite impuesto por la página que sirve para evitar estrangular los recursos del servidor donde viven los datos. El limite en Socrata es de 50000, más que suficiente para pedir todos esos datos pero tenemos que decirle, ¿Cómo? De esta manera:



In [7]:
source = "https://www.datos.gov.co/resource/4wwa-qb9a.json?$limit=10000"
info_carga = pd.read_json(source)
print(info_carga.shape)
info_carga.head()

(2584, 17)


fecha  ... carga_vuelos_internacionales
0  2020-03-25T00:00:00.000  ...                          353
1  2020-03-25T00:00:00.000  ...                            0
2  2020-03-25T00:00:00.000  ...                            0
3  2020-03-25T00:00:00.000  ...                            0
4  2020-03-25T00:00:00.000  ...                            0

[5 rows x 17 columns]

## Datos abiertos sobre COVID19 en Colombia

Para el trabajo con datos en tiempo real se necesita de un poco más que cambiar unas lienas en un link, no vamos a hondar en este tema (propio de desarrollo web), pero aquí tenemos un ejemplo de todo lo que se tendría que hacer para obtener un conjunto de datos online:

In [8]:
#@title Instalando en Colab
!pip install sodapy

In [0]:
#@title Esto es Python para programación web
import pandas as pd
from sodapy import Socrata
from datetime import date

def conectapp(dominio,dataset_id,token=None):
  with Socrata(dominio,token) as client:
    try:
      metadata = client.get_metadata(dataset_id)
    except Exception as e:
      print("Hay un error al iniciar el cliente o acceder a la base de datos.")
      print("Esta es la respuesta:",e)
    else:
      createdAt = date.fromtimestamp(metadata['createdAt'])
      updateAt = date.fromtimestamp(metadata['viewLastModified'])
      print("La base de datos fue creada en:",createdAt,",su ultima modificación está registrada en",updateAt)
    finally:
      return client,metadata

def sodatoDataFrame(dominio,dataset_id,token=None,limit = 5000):
  client,metadata = conectapp(dominio,dataset_id,token=token)
  nameFirstCol = metadata['columns'][0]['fieldName']
  rows_id = int(client.get(dataset_id, select="COUNT({0})".format(nameFirstCol))[0]['COUNT_{0}'.format(nameFirstCol)])
  print("El número de datos identificados es:", rows_id)
  isMoreData = rows_id > limit 
  results = client.get(dataset_id,limit=limit)
  if not isMoreData:
    dataset = pd.DataFrame.from_records(results)
  else:
    delta,start,i = limit,limit,0  
    dataset_list = [pd.DataFrame.from_records(results)]
    row_current = limit
    while isMoreData:
      start += delta*i
      results = client.get(dataset_id,limit=limit,offset=start)
      dataset_list.append( pd.DataFrame.from_records(results) )
      i +=1
      row_current += limit*i
      isMoreData = rows_id > row_current
    dataset = pd.concat(dataset_list,ignore_index=True)
  return dataset

In [14]:
dominio = "www.datos.gov.co"
#Este token (o llave, clave) será removido en los siguientes días, sin embargo
#usted puede obtener uno fácilmente o de plano no usarlo y poner None
#para más información https://dev.socrata.com/docs/app-tokens.html
token = "99wufepiZdd52go4LdzeENbG4"
dataset_id = "gt2j-8ykr"
dataset = sodatoDataFrame(dominio,dataset_id,token=token)

La base de datos fue creada en: 2020-03-27 ,su ultima modificación está registrada en 2020-05-09
El número de datos identificados es: 10051


In [15]:
dataset.head()

id_de_caso  ...        fecha_reporte_web
0          1  ...  2020-03-06T00:00:00.000
1          2  ...  2020-03-09T00:00:00.000
2          3  ...  2020-03-09T00:00:00.000
3          4  ...  2020-03-11T00:00:00.000
4          5  ...  2020-03-11T00:00:00.000

[5 rows x 16 columns]

##Consulta y limpiado de datos

Usualmente es natural encontrarse con bases de datos "contaminadas". Con esto se suele referir a que alguna colección puede estar mal almacenada, o bien puede que simplemente haya que cambiar de formato alguna serie. A continuación haremos uso de una base de datos obtenida del minsterio de salud colombiano sobre los casos nacionales de COVID-19 [Casos-positivos-de-COVID-19-en-Colombia](https://www.datos.gov.co/Salud-y-Protecci-n-Social/Casos-positivos-de-COVID-19-en-Colombia/gt2j-8ykr/data).

En la tabla encontraremos la información caso por caso. Veamoslo

In [16]:
import pandas as pd

covid_data= pd.read_csv('https://raw.githubusercontent.com/tomasate/Curso0/master/cvs_files/Casos_positivos_de_COVID-19_en_Colombia_old.csv')
columnas = covid_data.columns
covid_data.head()

ID de caso  ...        fecha reporte web
0           1  ...  2020-03-06T00:00:00.000
1           2  ...  2020-03-09T00:00:00.000
2           3  ...  2020-03-09T00:00:00.000
3           4  ...  2020-03-11T00:00:00.000
4           5  ...  2020-03-11T00:00:00.000

[5 rows x 16 columns]

O podemos usar el conjunto de datos actualizado a su fecha más reciente, como vimos en la sección anterior:

In [17]:
backup = covid_data.copy(deep=True)
covid_data = dataset.copy(deep=True)
covid_data.columns = columnas
typeint = ["ID de caso","Codigo DIVIPOLA","Edad"]
for t in typeint:
  covid_data[t]=covid_data[t].astype(int)
covid_data.head()

ID de caso  ...        fecha reporte web
0           1  ...  2020-03-06T00:00:00.000
1           2  ...  2020-03-09T00:00:00.000
2           3  ...  2020-03-09T00:00:00.000
3           4  ...  2020-03-11T00:00:00.000
4           5  ...  2020-03-11T00:00:00.000

[5 rows x 16 columns]

Como se puede ver tenemos bastante información sobre los casos: una etiqueta dispuesta por el ministerio para identificar cada caso, esta etiqueta podría servirnos de índice, luego veremos porqué. Tenemos una fecha de cuando la persona consultó. Un código asociado al municipio de residenia llamado DIVIPOLA, la Ciudad de ubicación, el Distrito o Departamento, el tipo de atención que está recibiendo por el virus, la Edad, el Sexo, una columna llamada Tipo que se refiere a si contrajo el virus en el pais o lo trajo desde el extranjero, una fecha aproximada sobre la cual pudo contraer el virus, la fecha de defunción, la fecha de diagnóstico, la fecha de recuperación y la fecha de reporte web. 

Por el momento entonces haremos una copia de este data frame, y trabajermos sobre la copia, esto con el motivo de que muchos procesos de limpiado de datos, como se puede esperar son procesos destructivos, por lo que nos interesa mantener una copia intacta a la que podamos retornar dada la necesidad. Además cambiaremos el nombre de las columnas con el fin de evitar caracteres que puedan complicar nuestra labor a la hora de manipular estos datos.

In [20]:
col_to_keep = ['ID de caso', 'Fecha de notificación',
               'Ciudad de ubicación', 'Departamento o Distrito ', 'atención', 'Edad',
               'Sexo', 'Tipo', 'Estado', 'País de procedencia', 'FIS',
               'Fecha de muerte', 'Fecha diagnostico', 'Fecha recuperado']

df = covid_data[col_to_keep]
df.columns = ['ID', 'Notificado',
               'Ciudad', 'Departamento_Distrito', 'Atencion', 'Edad',
               'Sexo', 'Tipo', 'Estado', 'Procedencia', 'Contagiado',
               'Muerte', 'Diagnosticado', 'Recuperado']

df.set_index('ID')

Notificado  ...               Recuperado
ID                              ...                         
1      2020-03-02T00:00:00.000  ...  2020-03-13T00:00:00.000
2      2020-03-06T00:00:00.000  ...  2020-03-19T00:00:00.000
3      2020-03-07T00:00:00.000  ...  2020-03-15T00:00:00.000
4      2020-03-09T00:00:00.000  ...  2020-03-26T00:00:00.000
5      2020-03-09T00:00:00.000  ...  2020-03-23T00:00:00.000
...                        ...  ...                      ...
10087  2020-05-08T00:00:00.000  ...                    -   -
10088  2020-05-08T00:00:00.000  ...                    -   -
10089  2020-05-08T00:00:00.000  ...                    -   -
10090  2020-05-08T00:00:00.000  ...                    -   -
10091  2020-05-08T00:00:00.000  ...                    -   -

[10051 rows x 13 columns]

Con lo anterior podemos notar la primera anomalía ¿Por qué hay más casos que filas? en principio debería haber tantos casos como filas, por lo que es intuitivo persar que hayan casos que no estén en la tabla, o bien los casos faltantes no han sido incluidos en la tabla. Aparte de las discusiones conspiracionistas, ya tenemos algo de información de la tabla: Está incompleta. Podemos hacer un par de ciclos que nos muestre los números faltantes.

In [19]:
for i in range(1,len(df)):
  if i != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, sume al siguiente loop', df['ID'][i-1]-i)
    break
for i in range(4156,len(df)):
  if i+14 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+14,'sume al siguiente loop', df['ID'][i-1]-i)
    break
for i in range(4160,len(df)):
  if i+15 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+15,'sume al siguiente loop', df['ID'][i-1]-i )
    break
for i in range(4390,len(df)):
  if i+22 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+22,'sume al siguiente loop', df['ID'][i-1]-i )
    break
for i in range(4430,len(df)):
  if i+25 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+25,'sume al siguiente loop', df['ID'][i-1]-i )
    break
for i in range(4571,len(df)):
  if i+39 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+39,'sume al siguiente loop', df['ID'][i-1]-i )
    break
for i in range(4827,len(df)):
  if i+40 != df['ID'][i-1]:
    print('El ', i, '-simo término corresponde a el ', df['ID'][i-1], ' ID, cuando debería ser', i+40,'sume al siguiente loop', df['ID'][i-1]-i )
    break

El  4156 -simo término corresponde a el  4170  ID, sume al siguiente loop 14
El  4160 -simo término corresponde a el  4175  ID, cuando debería ser 4174 sume al siguiente loop 15
El  4390 -simo término corresponde a el  4412  ID, cuando debería ser 4405 sume al siguiente loop 22
El  4430 -simo término corresponde a el  4455  ID, cuando debería ser 4452 sume al siguiente loop 25
El  4571 -simo término corresponde a el  4610  ID, cuando debería ser 4596 sume al siguiente loop 39
El  4827 -simo término corresponde a el  4867  ID, cuando debería ser 4866 sume al siguiente loop 40


In [22]:
df.iloc[4152:4158]

ID  ...               Recuperado
4152  4153  ...                    -   -
4153  4154  ...  2020-04-25T00:00:00.000
4154  4155  ...                    -   -
4155  4170  ...                    -   -
4156  4171  ...                    -   -
4157  4172  ...                    -   -

[6 rows x 14 columns]

Existe entonces un desfalco en 40 datos, sabemos además entonces que los ID desde el 4156 hasta 4169 no están identificados, lo mismo ocurre con el 4175, con los de 4405 hasta 4412 y así sucesivamente.

Podríamos interesarnos también en qué tipo de géneros está guardando el gobierno sobre los afectados por el virus, para esto podríamos usar el comando `.unique`

In [0]:
df['Sexo'].unique()

array(['F', 'M'], dtype=object)

Es posible que hayan otros datos con fallas de este estilo, haga una búsqueda de ello con este comando.
*edit: Actualizando podemos ver que es un error corregido*


In [0]:
backup['Sexo'].unique()

array(['F', 'M', 'm', 'f'], dtype=object)

Podríamos decirle a pandas que convierta los datos de fecha a un formato que Pandas trae por defecto. Intentemos hacerlo para dato de `Notificado`

In [0]:
df.Notificado = pd.to_datetime(df.Notificado)
df.head()

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


ID Notificado  ...            Diagnosticado               Recuperado
0   1 2020-03-02  ...  2020-03-06T00:00:00.000  2020-03-13T00:00:00.000
1   2 2020-03-06  ...  2020-03-09T00:00:00.000  2020-03-19T00:00:00.000
2   3 2020-03-07  ...  2020-03-09T00:00:00.000  2020-03-15T00:00:00.000
3   4 2020-03-09  ...  2020-03-11T00:00:00.000  2020-03-26T00:00:00.000
4   5 2020-03-09  ...  2020-03-11T00:00:00.000  2020-03-23T00:00:00.000

[5 rows x 14 columns]

Podemos ver que ha botado la parte de la fecha que databa de la hora (la cual no era nada útil, pues todas estaban llenas on `T00:00:00`) podríamos intentar hacer lo mismo para otras filas como la muerte 

In [0]:
try:
  df.Muerte = pd.to_datetime(df.Muerte)
except Exception as e:
  print("Error:\n",e)

Error:
 String does not contain a date: -   -


Tenemos un problema y es que hay un string que no podemos convertir a este tipo de formtos, cambiémoslo.

In [0]:
df = df.replace({df.Muerte[0]:None})
df.Muerte = pd.to_datetime(df.Muerte)
df

ID Notificado  ...            Diagnosticado               Recuperado
0          1 2020-03-02  ...  2020-03-06T00:00:00.000  2020-03-13T00:00:00.000
1          2 2020-03-06  ...  2020-03-09T00:00:00.000  2020-03-19T00:00:00.000
2          3 2020-03-07  ...  2020-03-09T00:00:00.000  2020-03-15T00:00:00.000
3          4 2020-03-09  ...  2020-03-11T00:00:00.000  2020-03-26T00:00:00.000
4          5 2020-03-09  ...  2020-03-11T00:00:00.000  2020-03-23T00:00:00.000
...      ...        ...  ...                      ...                      ...
10046  10087 2020-05-08  ...  2020-05-08T00:00:00.000                     None
10047  10088 2020-05-08  ...  2020-05-08T00:00:00.000                     None
10048  10089 2020-05-08  ...  2020-05-08T00:00:00.000                     None
10049  10090 2020-05-08  ...  2020-05-08T00:00:00.000                     None
10050  10091 2020-05-08  ...  2020-05-08T00:00:00.000                     None

[10051 rows x 14 columns]

Bien podemos hacer lo mismo para los otros items que contengan fechas

In [0]:
df.Estado = df.Estado.str.capitalize()
df.Diagnosticado = pd.to_datetime(df.Diagnosticado)
df = df.replace({df.Contagiado[25]:None})
df.Contagiado = pd.to_datetime(df.Contagiado)
df.Recuperado = pd.to_datetime(df.Recuperado)
df

ID Notificado               Ciudad  ... Muerte Diagnosticado  Recuperado
0          1 2020-03-02          Bogotá D.C.  ...    NaT    2020-03-06  2020-03-13
1          2 2020-03-06  Guadalajara de Buga  ...    NaT    2020-03-09  2020-03-19
2          3 2020-03-07             Medellín  ...    NaT    2020-03-09  2020-03-15
3          4 2020-03-09             Medellín  ...    NaT    2020-03-11  2020-03-26
4          5 2020-03-09             Medellín  ...    NaT    2020-03-11  2020-03-23
...      ...        ...                  ...  ...    ...           ...         ...
10046  10087 2020-05-08        Villavicencio  ...    NaT    2020-05-08         NaT
10047  10088 2020-05-08        Villavicencio  ...    NaT    2020-05-08         NaT
10048  10089 2020-05-08        Villavicencio  ...    NaT    2020-05-08         NaT
10049  10090 2020-05-08        Villavicencio  ...    NaT    2020-05-08         NaT
10050  10091 2020-05-08        Villavicencio  ...    NaT    2020-05-08         NaT

[10051 rows x 14 columns]

#**Oparaciónes de agregación y agrupamiento**

## **Introducción**

La categorización de un conjunto de datos y la aplicación de una función a cada grupo, ya sea una agregación o transformación, a menudo es un componente crítico de un flujo de trabajo de análisis de datos. Después de cargar y preparar un dataset, una tarea común es calcular estadísticas de grupo o tablas dinámicas para fines de informes o visualización. `pandas` proporciona una herramienta de agrupado flexible y de alto rendimiento, que le permite cortar, cortar en grupos y resumir conjuntos de datos de una manera natural.

Como veremos a continuación, con la funcionalidad y el poder de `Python` y `pandas`, podemos realizar agrupamientos complejos, utilizando cualquier cualquier función que acepte un objeto pandas o un arreglo `NumPy`. En este capítulo, aprenderá a:

##**Método `GroupBy`**

En lo subsecuente, trabajaremos con el concepto de **dividir-aplicar-combinar** para hablar sobre operaciones grupales. En la primera etapa del proceso, los datos contenidos en un objeto pandas, ya sea una Serie, un `DataFrame` o de otro modo, se dividen en grupos en función basados en una o más claves que se proporcionen. La división se realiza en un eje particular de un objeto. Por ejemplo, un `DataFrame` se puede agrupar en sus filas (`axis=0`) o sus columnas (`axis= 1`). Una vez hecho esto, se aplica una función a cada grupo, produciendo un nuevo valor. Finalmente, los resultados de todas esas aplicaciones de funciones se combinan en un objeto final. La forma del objeto resultante generalmente dependerá de lo que se haga con los datos. 

![texto alternativo](https://github.com/diplomado-bigdata-machinelearning-udea/curso0/blob/master/s07/imagenes/group_process.png?raw=true)

Cada clave de agrupamiento puede tomar muchas formas, y las claves no tienen que ser todas del mismo tipo:

- Una lista o arreglo de valores que es de la misma longitud que eje que se está agrupando.

- Un valor que indica el nombre de una columna en un `DataFrame`

- Un `dict` o una `Series` que proporciona una correspondencia entre los valores del eje y los nombres de los grupos.

- Una función que se invocará en el índice del eje o en las etiquetas individuales del índice.

Nótese que los últimos tres métodos son simplemente "atajos" para producir un arreglo de valores que será usado para dividir el objeto. A lo largo de la presenta sección iremos brindando ejemplos de estos métodos. 

In [0]:
df = pd.DataFrame({'key1' : ['a', 'a', 'b', 'b', 'a'],'key2' : ['uno', 'dos', 'uno', 'dos', 'uno'],
                'data1' : np.random.randn(5), 'data2' : np.random.randn(5)})

df

key1 key2     data1     data2
0    a  uno  0.710005 -1.552176
1    a  dos -0.651236 -0.040397
2    b  uno  0.569436  1.945155
3    b  dos -0.814546  0.387436
4    a  uno  0.934662  1.591392

Sumpongamos que se quiere calcular el valores medio de la columna `data1` usando los grupos de etiquetas `key1`. Existen muchas formas de hacer esto. Una de ellas, consiste en acceder a `data1` y aplicar el método `groupby` con la columna (de tipo `Series`) en la clave `key1`:

In [0]:
grouped = df['data1'].groupby(df['key1'])

grouped

La variable `grouped` se conoce como un objeto `GroupBy`. No se ha calculado nada aun, a excepción de de unos datos intemedios sobre laa clave de grupo `df['key1']`. La idea es que este objeto tenga toa la información necesaria para luego aplicar alguna operación a cada uno de los grupos. Por ejemplo, para calcular la media por grupo, podemos llamar el método `mean()` del objeto `GroupBy`:

In [0]:
grouped.mean()

key1
a    0.331144
b   -0.122555
Name: data1, dtype: float64

Nóteses que, lo importante aquí es que los datos (un objeto `Series`) han sido agrupados de acuerdo con los valores de la clave 'key1'. Se hace evidente, además, que el índice del objeto `Series` resultante tiene el nombre `key1` ya que se realizo mediante la especificación de `df['key1']`.


Veamos que, efectivamente, el tipo de dato devuelve el método que acabamos de aplicar es `Series`:

In [0]:
type(grouped.mean())

pandas.core.series.Series

Si pasamos una lista de varios arreglos, el resultado es diferente:

In [0]:
means = df['data1'].groupby([df['key1'], df['key2']]).mean()
means

key1  key2
a     dos    -0.651236
      uno     0.822334
b     dos    -0.814546
      uno     0.569436
Name: data1, dtype: float64

En este caso, hemos agrupado los datos usando dos claves. Se evidencia que objeto `Series` posee una jerarquización en el indexado de las parejas únicas de las claves observadas.

Usemos el método `unstack()`, para hacer una combinación de todas las claves y visualizar los valores medios en cada caso:

In [0]:
means.unstack()

key2       dos       uno
key1                    
a    -0.651236  0.822334
b    -0.814546  0.569436

En estos ejemplos, las claves son todas `Series`; téngase en cuenta que pueden ser cualquier array de la misma longitud de la tabla aprpropiada 

In [0]:
len(df)

5

Si usamos otra serie, o arreglo, con una longitud de cinco elementos, podremos agrupar nuestro `DataFrame` `df`:

In [0]:
departamentos = np.array(['Antioquia', 'Choco', 'Cundinamarca', 'Amazonas', 'Vaupes'])
tiempo = np.array([2008, 2008, 2009, 2008, 2009])

df['data1'].groupby([departamentos, tiempo]).mean()

Amazonas      2008   -0.814546
Antioquia     2008    0.710005
Choco         2008   -0.651236
Cundinamarca  2009    0.569436
Vaupes        2009    0.934662
Name: data1, dtype: float64

Con frecuencia, la información de agrupación se encuentra en el mismo DataFrame que los datos en los que desea trabajar. En ese caso, puede pasar los nombres de columna (ya sean cadenas, números u otros objetos de Python) como las claves del grupo:

In [0]:
df.groupby('key1').mean()

data1     data2
key1                    
a     0.331144 -0.000393
b    -0.122555  1.166295

In [0]:
 df.groupby(['key1', 'key2']).mean()

data1     data2
key1 key2                    
a    dos  -0.651236 -0.040397
     uno   0.822334  0.019608
b    dos  -0.814546  0.387436
     uno   0.569436  1.945155

Nótese que en el primer caso `def.groupby('key1').mean()` no contiene columna `key2` en su resultado. Lo anterior se debe a que `df['key2']` no es de tipo numérico, decimos que es una columna 'problemática', y por lo tanto se excluye del resultado. De forma predeterminada, todas las columnas numéricas son agregadas, aunque es posiblre filtrarlas a un subconjunto menor, como se verá más adelante.

Independientemente del objetivo al usar `groupby`, un método `GroupBy` generalmente útil corresponde al tamaño que retorna una serie que contiene tamaños de los grupo:

In [0]:
df.groupby(['key1', 'key2']).size()

key1  key2
a     dos     1
      uno     2
b     dos     1
      uno     1
dtype: int64

##**Iteración sobre Grupos:**

El objeto `GroupBy` admite la iteración, generando una secuencia de 2 tuplas que contienen el nombre del grupo junto con la porción de datos. Considere el siguiente conjunto de datos de ejemplo:

In [0]:
for nombre, grupo in df.groupby('key1'):
  print("Nombre:", nombre)
  print(grupo)


Nombre: a
  key1 key2     data1     data2
0    a  uno  0.710005 -1.552176
1    a  dos -0.651236 -0.040397
4    a  uno  0.934662  1.591392
Nombre: b
  key1 key2     data1     data2
2    b  uno  0.569436  1.945155
3    b  dos -0.814546  0.387436


En el caso de claves múltiples, el primer elemento de la tupla será una tupla de los valores de las claves:

In [0]:
for (k1, k2), group in df.groupby(['key1', 'key2']):
  print("Claves: ",k1,',', k2)
  print(group)

Claves:  a , dos
  key1 key2     data1     data2
1    a  dos -0.651236 -0.040397
Claves:  a , uno
  key1 key2     data1     data2
0    a  uno  0.710005 -1.552176
4    a  uno  0.934662  1.591392
Claves:  b , dos
  key1 key2     data1     data2
3    b  dos -0.814546  0.387436
Claves:  b , uno
  key1 key2     data1     data2
2    b  uno  0.569436  1.945155


En general, se puede realizar cualquier tipo de operación con los datos. Una aplicación que puede resultar útil, es el cómputo de un `dict` de las partes de un conjunto de datos como una línea

In [0]:
dict(list(df.groupby('key1')))

{'a':   key1 key2     data1     data2
 0    a  uno  0.710005 -1.552176
 1    a  dos -0.651236 -0.040397
 4    a  uno  0.934662  1.591392, 'b':   key1 key2     data1     data2
 2    b  uno  0.569436  1.945155
 3    b  dos -0.814546  0.387436}

In [0]:
partes = dict(list(df.groupby('key1')))

In [0]:
partes['a']

key1 key2     data1     data2
0    a  uno  0.710005 -1.552176
1    a  dos -0.651236 -0.040397
4    a  uno  0.934662  1.591392

In [0]:
partes['b']

key1 key2     data1     data2
2    b  uno  0.569436  1.945155
3    b  dos -0.814546  0.387436

Por defecto, `groupby` agrupa a lo largo de `axis=0`, per puede agruparse en cualquier eje. Ppor jemplo, podríamos agrupas las columnas de nuestro ejemplo de acuerdo a los tipos de datos `dtypes` que almacena:

In [0]:
df.dtypes

key1      object
key2      object
data1    float64
data2    float64
dtype: object

In [0]:
grouped = df.groupby(df.dtypes, axis=1)
dict(list(grouped))

{dtype('float64'):       data1     data2
 0  0.710005 -1.552176
 1 -0.651236 -0.040397
 2  0.569436  1.945155
 3 -0.814546  0.387436
 4  0.934662  1.591392, dtype('O'):   key1 key2
 0    a  uno
 1    a  dos
 2    b  uno
 3    b  dos
 4    a  uno}

## **Selección de una Columna o subconjunto de columnas**

El indexado de un objeto `GroupBy`, el cual es creado a partir de un `DataFrame` con nombres de columnas o arreglos de nombres de columnas tiene el efecto de seleccionar esas columnas para la agregación. Lo anterior sigifica que



```
df.groupby('key1')['data1']
df.groupby('key1')[['data2']]
```
Es una sintaxis 'más' intuitiva que:

```
df[['data1']].groupby(df['key1'])
df[['data2']].groupby(df['key1'])
```

la cual es aceptable también.

En específico, para grandes conjuntos de datos, puede resultar conveniente agregar solo unas pocas columnas. Por ejemplo, en el conjunto de datos anterior, para calcular los valores medios solo de las columnas `data2` y obtener el resultado como un `DataFrame`, escribimos:


In [0]:
df.groupby(['key1', 'key2'])[['data2']].mean()

data2
key1 key2          
a    dos  -0.040397
     uno   0.019608
b    dos   0.387436
     uno   1.945155

El objeto que se devuelve en esta operación de indexación es un `DataFrame` agrupado siempre que se pase una lista o arreglo junto a un método como `mean` y una Serie agrupada si solo se pasa un nombre de la columna o lista de ellas:

In [0]:
 series_grouped = df.groupby(['key1', 'key2'])['data2']
 series_grouped

In [0]:
series_grouped.mean()

key1  key2
a     dos    -0.040397
      uno     0.019608
b     dos     0.387436
      uno     1.945155
Name: data2, dtype: float64

##**Agrupamiento con Diccionarios y Series**

La información del agrupamiento puede darse en una forma diferente de un arregle. Consideremos otro ejemplo de un DataFrame:

In [0]:
personas = pd.DataFrame(np.random.randn(5, 5), columns=['a', 'b', 'c', 'd', 'e'], index=['Jesus', 'Eesteban', 'William', 'Jaime', 'Tomas'])
personas

a         b         c         d         e
Jesus    -0.508605  0.372129 -0.823437  0.710051 -0.597236
Eesteban  1.400553  1.056252 -1.199701  1.535257  0.770771
William  -0.252152  1.233554  1.368270 -0.537815  0.227810
Jaime    -0.159103  0.611904 -0.480612  1.521181  0.285538
Tomas     0.333522  0.560947 -0.123565 -1.066583 -0.737061

Ahora, supongamos que tengo una correspondencia grupal para las columnas y deseo sumar las columnas por grupo:

In [0]:
 mapa = {'a': 'rojo', 'b': 'rojo', 'c': 'azul','d': 'azul', 'e': 'rojo', 'f' : 'naranja'}

A partir de este mapa, podriamos construir fácilmente un arreglo pasándoselo al método `groupby`, por ejemplo, para hacer una suma a lo largo del eje de las columnas:

In [0]:
 por_columnas = personas.groupby(mapa, axis=1)
 por_columnas.sum()

azul      rojo
Jesus    -0.113387 -0.733712
Eesteban  0.335556  3.227576
William   0.830455  1.209212
Jaime     1.040569  0.738338
Tomas    -1.190148  0.157408

o realizar una cuenta

In [0]:
personas.groupby(mapa, axis=1).count()

azul  rojo
Jesus        2     3
Eesteban     2     3
William      2     3
Jaime        2     3
Tomas        2     3

##**Agrupamiento con funciones**

Cualquier función que se pase como clave se llamará una vez por valor de índice, y los valores devueltos se utilizarán como nombres de grupo. En concreto, consideremos el `DataFrame` anterior, cuyo índice corresponde a nombres de personas. Supongamos que queremos agrupar por longitud de los nómbres; se ppodría calcular un areglo de longitudes de los nombres, pero en su lugar, le podemos pasar al método `groupby` la función `len`:

In [0]:
personas.groupby(len).sum()

a         b         c         d         e
5 -0.334187  1.544981 -1.427614  1.164649 -1.048760
7 -0.252152  1.233554  1.368270 -0.537815  0.227810
8  1.400553  1.056252 -1.199701  1.535257  0.770771

Si realizamos una mezcla de funciones con arreglos, diccionarios o series, todo es convertido internamente en un arreglo:

In [0]:
lista_claves = ['uno', 'uno', 'uno', 'dos', 'dos']
personas.groupby([len,lista_claves]).min()

a         b         c         d         e
5 dos -0.159103  0.560947 -0.480612 -1.066583 -0.737061
  uno -0.508605  0.372129 -0.823437  0.710051 -0.597236
7 uno -0.252152  1.233554  1.368270 -0.537815  0.227810
8 uno  1.400553  1.056252 -1.199701  1.535257  0.770771

##**Agrupamiento por niveles de índice o indexado múltiple**

Nótese, que en alguno de los ejemplos desarollados, los `DataFrames` presentan dos índices. En el caso de presentar dos más índices, hablamos de un **indexado múltiple**. Implicitamente, se establece una jerarquía decreciente, de izquierda a derecha, en el indexado.

De forma conveniente, podemos usar esta jerarquización para agrupar los datos usando uno de los niveles de un eje de un índice. Para realizar esto, se pasa el número de nivel o el nombre, usando la palabra clave `level`. Vemos esto con un ejemplo.

In [0]:
columnas = pd.MultiIndex.from_arrays([['COL', 'COL', 'COL', 'VEN', 'VEN'],[1, 3, 5, 1, 3]], names=['pais', 'mes'])

In [0]:
hDF = pd.DataFrame(np.random.randn(4, 5), columns=columnas)
hDF

pais       COL                           VEN          
mes          1         3         5         1         3
0    -0.369991  1.199524  0.288103  0.520182  0.960170
1     0.500928 -0.274234 -0.712364 -0.643371  1.029284
2    -0.193242 -0.139353  0.543040  0.878083 -0.316152
3     0.489508 -2.107018 -0.072691 -0.569228  0.728282

Por ejemplo, podemos hacer un conteo basados en el pais, y a lo largo de las columnas del indexado múltiple:

In [0]:
hDF.groupby(level='pais',axis=1).count()

pais  COL  VEN
0       3    2
1       3    2
2       3    2
3       3    2

##Ejercicio 1.
Realice un estudio sobre el tiempo que perdura la enfermedad en un paciente, para ello intente en primer lugar crear una columna llamada `Recuperación` que será la diferencia en tiempo de `Recuperado` y `Contagiado`. Una vez lo haga intente responder la pregunta ¿Qué pasa con los datos de `Recuperado` que ahora son `NaT`? Hecho esto, utilice la operación `groupby` para agrupar los datos en función de `Recuperación` y saber cuántos datos tardaron cuántos días en recuperarse. 

Saque un promedio ponderado de los días que tarda una persona en recuperarse.

Puede ver la solución **_aquí_**
<!---
df1 = df.copy()
df1['Recuperacion'] = df['Recuperado']-df['Contagiado']
s = df1.groupby('Recuperacion')['Edad'].count()
td = s.index
suma = td[0]-td[0]
perso = 0
for i in range(len(td)):
  suma = td[i]*s[i]+suma
  perso = s[i]+ perso
x = suma/perso
days = x.astype('timedelta64[D]')
days
--->

##Ejercicio 2.

Realice un estudio que permita determinar el tiempo de respuesta que se tuvo con el virus a partir de la diferencias de fechas de `Diagnosticado` y `Contagiado`. 

Consulte la solución **_aquí_**
<!---
df2 = df.copy()
df2['tiempo_en_diagn'] = df['Diagnosticado']-df['Contagiado']
s = df2.groupby(['tiempo_en_diagn'])['Edad'].count()
#s = df1.groupby('Recuperacion')['Edad'].count()
td = s.index
suma = td[0]-td[0]
perso = 0
for i in range(len(td)):
  suma = td[i]*s[i]+suma
  perso = s[i]+ perso
x = suma/perso
days = x.astype('timedelta64[D]')
days

--->

##Ejercicio 3.

Consulte cuál es el departamento o distrito con más contagiados, una vez haga esto, cree un nuevo dataframe que solo contenga la información de este departamento o distrito y posterior a esto agrupe los datos según el género de este.

Consulte la solución **_aquí_**
<!---
s = df.groupby(['Departamento_Distrito'])['Edad'].count()
maxi =max(s)
elmax = s.where(s == maxi).dropna()
Depto = elmax.index[0]
Depto
data_bogota = df.where(df['Ciudad'] == Depto)
data_bogota.groupby('Sexo')['Edad'].count()
--->